Draft: Opt-In Continuous Delivery of Jenkins Plugin Releases (#244)
jenkinsci/jep: jep/0000/README.adoc

* Trusted CI/CD: deployments are produced from a single source of truth, which is managed by 
the Jenkins project

== Specification

=== Overview

We'll use a short story to illustrate what it will look like when a plugin maintainer 
named Karl decides that he wants to perform continuous delivery of his (fictional) `karl-plugin`.

To do so, Karl will:

1. Verify support. Karl's plugin is already in the 
link:https://github.com/jenkinsci[jenkinsci Organization on GitHub], and is 
using `buildPlugin` when built on ci.jenkins.io.

2. Choose a branch from which to release. Karl decides he wants to release from the `master` branch, 
as that is the simplest.

3. Settle on a version numbering scheme. Becuase it's in fairly wide use across the Jenkins 
ecosystem, Karl settles on link:https://semver.org/[semantic versioning].

4. Add a file to the top directory of his repository called `Autorelease.yaml`. 

5. A webhook is configured on Karl's repository on GitHub to point to a webhook receiver. 

Once ready to create a new release of his plugin, Karl merges some changes to the `master` branch. 
At which point, the following takes place under the covers: 

> 1. A webhook is sent to the webhook receier by GitHub

Reviewer: ```suggestion
1. A webhook is sent to the webhook receiver by GitHub
```

Author: Good catch. Writing English with a code editor is...well, it makes one appreciate a word processor. :)

//| Resolution
//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:

|===

== Abstract

This JEP introduces a new system which enables Jenkins plugin maintainers to add Continuous Delivery (CD) to their projects.
It is:

* Opt-in: plugin maintainers can choose to use this
* Simple to configure: very little needs to be changed in order to enable CD
* Trusted CI/CD: deployments are produced from a single source of truth, which is managed by 
the Jenkins project

== Specification

=== Overview

We'll use a short story to illustrate what it will look like when a plugin maintainer 
named Karl decides that he wants to perform continuous delivery of his (fictional) `karl-plugin`.

To do so, Karl will:

1. Verify support. Karl's plugin is already in the 
link:https://github.com/jenkinsci[jenkinsci Organization on GitHub], and is 
using `buildPlugin` when built on ci.jenkins.io.

2. Choose a branch from which to release. Karl decides he wants to release from the `master` branch, 
as that is the simplest.

3. Settle on a version numbering scheme. Karl decides on link:https://semver.org/[semantic versioning].

> 4. Add a file to the top directory of his repository called `Autorelease.yaml`. 

Reviewer: Couldn't this be done as an argument to `buildPlugin()`?

Author: Noted as something to come back to in the latest revision.

    </licenses>

    <properties>
        <revision>3.1.4</revision>
        <changelist>-SNAPSHOT</changelist>
        <java.level>8</java.level>
        <jenkins.version>2.138.4</jenkins.version>
    </properties>

----

No additional changes need to be made to `pom.xml` by Karl, he's good to go.

==== Autorelease.yaml verification in `buildPlugin`

The commonly used library link:https://github.com/jenkins-infra/pipeline-library/blob/master/vars/buildPlugin.groovy[`buildPlugin`] will need to be modified to check for the presence of, and validate, `Autorelease.yaml`. This 
markerfile will be the mechanism that tells the trusted CI server that this plugin should be automatically 
released. 

Validation must include, but will not be limited to, the following:

1. `pluginName` field must match the repository name

2. `versionNumberFields` must match the `revision` property in `pom.xml`. E.g., if `versionNumberFields` is 
set to `3` (three fields), and the `revision` in `pom.xml` is set to `3.1.4.5` (four fields), validation 
will fail.

3. Validation that no extra fields are present in the file. Comments are allowed, but any unexpected 
lines will cause validation to fail.

> If validation of `Autorelease.yaml` fails for any reason, the build is not performed nothing gets deployed, 

Reviewer: ```suggestion
If validation of `Autorelease.yaml` fails for any reason, the build is not performed, nothing gets deployed, 
```

= JEP-0000: Opt-in Continuous Delivery of Jenkins Plugin Releases
:toc: preamble
:toclevels: 3
ifdef::env-github[]
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
endif::[]

.Metadata
[cols="1h,1"]
|===
| JEP
| 0000

| Title
| Opt-in Continuous Delivery of Jenkins Plugin Releases from Trusted CI

| Sponsor

> | link:https://github.com/daniel-beck[Daniel Beck]

Reviewer 2: I might need to check what I'm signing up for, but sure 😅  I want to see this happen!

> | link:https://github.com/daniel-beck[Daniel Beck]

Author: I apologize if I misheard you Daniel, but you're stuck with it now. :stuck_out_tongue:

// Use the script `set-jep-status <jep-number> <status>` to update the status.
| Status
| Draft :speech_balloon:

| Type
| Process :bulb:

| Created
| Date (2019-05-15) :bulb:

| BDFL-Delegate
| TBD

//
//
// Uncomment if there is an associated placeholder JIRA issue.
//| JIRA
//| :bulb: https://issues.jenkins-ci.org/browse/JENKINS-nnnnn[JENKINS-nnnnn] :bulb:
//
//
// Uncomment if discussion will occur in forum other than jenkinsci-dev@ mailing list.
//| Discussions-To
//| :bulb: Link to where discussion and final status announcement will occur :bulb:
//
//
// Uncomment if this JEP depends on one or more other JEPs.
//| Requires
//| :bulb: JEP-NUMBER, JEP-NUMBER... :bulb:
//
//
// Uncomment and fill if this JEP is rendered obsolete by a later JEP
//| Superseded-By
//| :bulb: JEP-NUMBER :bulb:
//
//
// Uncomment when this JEP status is set to Accepted, Rejected or Withdrawn.
//| Resolution
//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:

|===

== Abstract

This JEP introduces a new system, which enables Jenkins plugin maintainers to add Continuous 

> Delivery to their projects. It is:

Author: +1 to that. It's tricky to imagine what this will look like in the end.

* Opt-in: plugin maintainers can choose to use this
* Simple to configure: very little needs to be changed in order to enable CD

> * Trusted CI/CD: deployments are produced from a single source of truth, which is managed by 

Author: Great observation.

the Jenkins project

== Specification

=== Overview

> We'll use a short story to illustrate what Autorelease will look like when a plugin maintainer 

Author: Is it a reasonable term to use? Maybe I should go with something more generic, like;

> We'll use a short story to illustrate what continuous delivery will look like when a plugin maintainer

The config file is called `Autorelease.yaml`. 

I'm probably overthinking this.

> We'll use a short story to illustrate what Autorelease will look like when a plugin maintainer 

Author: I'm going with 

> We'll use a short story to illustrate what it will look like when a plugin maintainer...

> We'll use a short story to illustrate what Autorelease will look like when a plugin maintainer 

Reviewer 2: I think the point was simply that there's no sentence like "The new system described in this JEP will be called Autorelease" or similar.

named Karl decides that he wants to perform continuous delivery of his (fictional) Karl-Plugin.

To do so, Karl will:

1. Verify support. Karl's plugin is already in the 
link:https://github.com/jenkinsci[jenkinsci Organization on GitHub], and is 
using `buildPlugin` when built on ci.jenkins.io.

2. Choose a branch from which to release. Karl decides he wants to release from the `master` branch, 
as that is the simplest.

> 3. Settle on a version numbering scheme. Becuase it's in fairly wide use across the Jenkins 

Author: Yikes. I've done a lot of that kind of stuff here.

> ecosystem, Karl settles on link:https://semver.org/[semantic versioning].

Author: Karl is going to use it. :P  (and he also made a minor edit to this line to reflect reality)

4. Add a file to the top directory of his repository called `Autorelease.yaml`. 

> 5. A webhook is configured on Karl's repository on GitHub to point to `{the-webhook-receiver}`

Author: 1. The webhook receiver is a computer, which does not exist yet. It'll act as a proxy of sorts, to take these incoming webhooks from commits, and send them to a secure CI server which lives on a VPN. This is in order to keep the CI server in a safe place.

2. My original thought was that the plugin maintainer would configure the webhook on a per-repository basis. That way, if Jane Java decides she wants to use CD on her `jane-plugin`, but release her `old-timey-plugin` from her laptop like she's always done, she can do so.

Maybe an organization-level webhook is the right approach. I hadn't thought of it.  Is there any risk of the webhook receiver being overrun with a huge influx of webhooks, every time someone commits to a repository in the jenkinsci organization? Not necessarily from the standpoint of overwhelming that receiver, but rather, from a logging standpoint. Let's say something goes wrong in this system. The infra team might have an easier time troubleshooting if the only webhooks getting sent are from individual repos, rather than from every repo in the org.

> 5. A webhook is configured on Karl's repository on GitHub to point to `{the-webhook-receiver}`

Reviewer 3: > 1. The webhook receiver is…
> 2. My original thought was…

So write all this in the doc!

> Maybe an organization-level webhook is the right approach…

→ **Reasoning** section

> 5. A webhook is configured on Karl's repository on GitHub to point to `{the-webhook-receiver}`

Author: Done and done.

Once ready to create a new release of his plugin, Karl merges some changes to the `master` branch. 
At which point, the following takes place under the covers: 

1. A webhook is sent to `{the-webhook-receiver}` by GitHub

> 2. The receiver forwards this hook to `${trusted-ci-server}`, which is protected by a VPN

Reviewer 3: Again, `${trusted-ci-server}` is confusing syntax. Is this supposed to be an actual code variable? Or do you just mean to say

> a trusted CI server

where the details will be spelled out in another section?

Author: Fair question. I was using that representation to imply that "trusted-ci-server" was a thing we knew we needed, but didn't have yet. I was probably being overly cutesy there. :) 

If there's a better, more traditional way of implying that, I'm happy to do it that way instead. Perhaps, as you suggest, I could use 

a trusted CI server <sup>2</sup>

And have <sup>2</sup> link to a place in the document where I explain this briefly, and point out that further details will appear in the yet-to-be-written IEP?

Reviewer 4: +1 to Jesse's suggestion. 
I think for this section, you can use plain english and use internal linking to sections that describe these in more detail and link out to other docs if needed. (See https://github.com/jenkinsci/jep/blob/master/jep/1/README.adoc#overview for example of what I mean).

3. `${trusted-ci-server}` uses the enhanced version of `buildPlugin` to validate the correctness of 
Karl's `Autorelease.yaml` file, and the consistency of Karl's chosen version numbering.

4. Once `Autorelease.yaml` has been validated, a plugin build is performed on `${trusted-ci-server}` 
as one would expect.

5. After the build is successful, `${trusted-ci-server}` handles deployment of the built plugin to 

> Artifactory and the Jenkins Update Center.

Reviewer 3: To both? Or to Artifactory, which then causes the plugin to appear on the UC by virtue of the existing UC generator?

Author: Then I suppose to Artifactory. I didn't know that's how the UC gets populated. Will re-word to reflect that. Maybe 

> After the build is successful, the trusted CI server handles deployment of the built plugin to 
> Artifactory, at which point the existing UC generator makes the plugin available on the update center

If there's documentation on how that UC generator works, I'd (a) love to read that, and (b) be happy to have the words _existing UC generator_ link to that document.

Reviewer 2: https://github.com/jenkins-infra/backend-update-center2 is the tool for it. We build and run it basically all the time. It downloads the periodically updated Maven index from Artifactory, and uses that to build up a list of plugins and their releases to deliver. There's a bunch of additional features like overrides and performance optimizations here, so it might look pretty weird. That it uses a really obsolete library to do the Maven index stuff doesn't help.

Basically, in terms of deployment, nothing needs to be different, server-side, from today when we get artifacts into Artifactory, the infra knows how to handle it. (Modulo the variable length version numbering scheme we talked about.)

Reviewer 3: > nothing needs to be different

That we know of now.

Author: For what it's worth, I've added a link to the UC generator in the appropriate parts of Overview and Detailed Setup. If we find something that needs to be changed, we'll take that as it comes.

=== Detailed Setup

==== Autorelease.yaml

A YAML markerfile, stored in the plugin's own repository. Absence of this markerfile will 
be functionally equivalent to the plugin maintainer opting out of Autorelease. 

In the following example, we'll stick with plugin developer Karl, and his fictional 
plugin `karl-plugin`. Karl's `Autorelease.yaml` file looks like this:

```
plugin: karl
kind: plugin
metadata:
  autoreleaseEnabled: true
  triggeringBranch: master
  versionNumberFields: 3
```

> In the above example `Autorelease.yaml`, we can see the following:

Reviewer 3: [Callouts](https://www.methods.co.nz/asciidoc/chunked/ch20.html) would help here.

Author: That's great. Didn't know that was a thing. :1st_place_medal:

Author: This might need some tweaking. It's doing "something" but it's not as pretty looking as I was expecting it to be:

![image](https://user-images.githubusercontent.com/21689198/59532884-f14c9c80-8eb7-11e9-841c-477e4e357cd6.png)

Reviewer 3: I have no idea offhand if GitHub’s renderer supports them.

Author: It's still a pretty neat thing, though, thanks for showing that to me.

`plugin: karl`:: 
This is the name of the plugin, and it matches the repository 

> name on GitHub.

Reviewer 3: Well, no, it matches the `artifactId` I suppose, where the repository name is `${artifactId}-plugin`.

And in such a case, why specify it at all? At best it is going to be a source of typos and mistakes; at worst a security risk.

Author: Good call. I've stricken it.

`kind`:: 
The type of project which is being set up for autorelease. In our first iteration of 
autorelease, this will only be set to `plugin`. The field is here should the Jenkins project 
want to expand the use of autorelease, such as for building core, or components.

> `metadata`:: 

Reviewer 3: Why bother with the subsection? This is all “metadata”.

Contains the following fields:
** `autoreleaseEnabled`: A boolean value, indicating whether or not this plugin is participating 

> in autorelease. Allows the maintainer to turn Autorelease on or off, without deleting the file 

Reviewer 3: IMHO drop this field. We can comfortably delete files, because we have version control.

Author: Fair enough, I'll strike it.

entirely.
** `triggeringBranch`: Sets the branch which maintainers will merge to in order to trigger 
an autorelease webhook. Although this is expected to usually be set to `master`, plugin 
maintainers can choose a branch name of their preference, e.g., `autorelease`, `release`, 
etc.
** `versionNumberFields`: An integer, indicating the number of fields which `buildPlugin` should expect.
This is important, because sticking to a consistent numbering scheme makes for easier sorting by version.
For this example plugin, Karl has set the value to `3`, because he's usuing semantic versioning, e.g., 
3.1.4.

==== Maven POM file changes

> Karl's plugin is already making use of Incrementals, and the relevant lines of his `pom.xml` file 

Reviewer 3: ```suggestion
Karl's plugin is already making use of link:https://github.com/jenkinsci/incrementals-tools/blob/master/README.md#usage-in-plugin-poms[Incrementals], and the relevant lines of his `pom.xml` file 
```

look like:

> ```

Reviewer 3: ```suggestion
[source,xml]
----
```

for highlighting (terminate with another `----`).

<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <parent>
        <groupId>org.jenkins-ci.plugins</groupId>
        <artifactId>plugin</artifactId>
        <version>4.56</version>
        <relativePath />
    </parent>
    <artifactId>karl</artifactId>
    <version>${revision}${changelist}</version>
    <packaging>hpi</packaging>
    <name>Karl Plugin</name>
    <url>
        <!--Something like https://wiki.jenkins-ci.org/display/JENKINS/Karl+Plugin-->
    </url>
    <description>A useful description of the Karl plugin.</description>
    <licenses>
        <license>
            <name>MIT</name>
            <url>http://opensource.org/licenses/MIT</url>
        </license>
    </licenses>

    <properties>
        <revision>3.1.4</revision>
        <changelist>-SNAPSHOT</changelist>
        <java.level>8</java.level>
        <jenkins.version>2.138.4</jenkins.version>
    </properties>

> ```

Author: Will need to replace ``` with `----`

No additional changes need to be made to `pom.xml` by Karl, he's good to go.

==== Autorelease.yaml verification in `buildPlugin`

The commonly used library link:https://github.com/jenkins-infra/pipeline-library/blob/master/vars/buildPlugin.groovy[`buildPlugin`] will need to be modified to check for the presence of, and validate, `Autorelease.yaml`. This 
markerfile will be the mechanism that tells `${trusted-ci-server}` that this plugin should be automatically 
released. 

Validation must include, but will not be limited to, the following:

1. `pluginName` field must match the repository name

2. `versionNumberFields` must match the `revision` property in `pom.xml`. E.g., if `versionNumberFields` is 
set to `3` (three fields), and the `revision` in `pom.xml` is set to `3.1.4.5` (four fields), validation 

> will fail.

Reviewer 3: So…why require it to be specified to begin with?

Author: This is related to something that @daniel-beck and I talked about. I understood it in the moment, but the passage of time means I've lost the plot a little. We wanted the number of fields to remain consistent within a given plugin, to avoid having versions sort incorrectly.

Reviewer 2: If the pom declares a version prefix that is later amended by the tool, say, `3.1` or `3.1.${build}` or equivalent, we may run into the problem that a "later looking" version number is actually sorting as earlier.

Imagine the fixed version prefix `3.1`, and autorelease adding a build number:

3.1.0
3.1.1
3.1.2
3.1.3

Now, the maintainer changes the apparent semver to `3.1.1`.

3.1.1.0 (or 3.1.1.4, doesn't matter)
3.1.1.1
…

While these releases should be later, they sort earlier. Fixed length version numbers (or "consistent length" per component) addresses this problem.

I also noted this might be circumventable by using a different scheme here – E.g. if something like `3.1.1-b4` or similar would sort differently.

Reviewer 3: > Fixed length version numbers addresses this problem.

Until the maintainer

```diff
-versionNumberFields: 2
+versionNumberFields: 3
```

to fix a build error, without considering the consequences on the update center.

> something like `3.1.1-b4` or similar would sort differently

As per [JEP-305](https://github.com/jenkinsci/jep/blob/master/jep/305/README.adoc#other-version-schemes), which is not susceptible to this problem because the generated versions always sort _before_ the maintainer-specified string. Of course that particular format is undesirable in this context since these are not “release candidates”, these are full releases—there is no other.

----

My suggestion would be to just get rid of the maintainer-specified component altogether in an autoreleased plugin. It serves no function beyond [pretending to indicate compatibility](https://twitter.com/tyvole/status/1136767636602327042), which is more helpfully and honestly accomplished by writing good release notes. With a minor enhancement to `git-changelist-maven-extension` to allow the `-rc` component to be skipped, the POM can just say

```xml
<version>${changelist}</version>
<properties>
  <changelist>999999-SNAPSHOT</changelist>
</properties>
```

so development builds are all `999999-SNAPSHOT` and incremental builds (including `master` releases) look like `1234.abc123def456`. The more changes you make, the bigger the number, and you never have to pay attention to the version again.

Reviewer 3: …in which case we do not need another server, or even a new YAML file. ci.jenkins.io will already publish these releases. We would “just” need to patch `update-center2` to also check the Incrementals repository and include the newest release it finds there which

* Matches `\d+[.][a-f0-9]{12}` (i.e., you have opted in by virtue of dropping the manual `revision`).
* Comes from `master` according to [the same logic already used by `incrementals:update`](https://github.com/jenkinsci/incrementals-tools/blob/9a96211ad69fa55599001c9689b5c7e78f9a7f88/lib/src/main/java/io/jenkins/tools/incrementals/lib/UpdateChecker.java#L141-L146).

Reviewer 3: > patch `update-center2` to also check the Incrementals repository

Better yet, make the `incrementals-publisher` function send artifacts matching these criteria to the regular `releases` repository; or have a separate function or batch process which copies from `incrementals` to `releases`. Thus

* These releases may be made dependencies of other plugins without `-Pconsume-incrementals`.
* We can experiment with GC policies in `incrementals` without fear of dropping bits which were at some point offered on the regular user UC.
* `update-center2` need not be touched—these artifacts would appear in the `releases` index and be published automatically.

Reviewer 2: > to fix a build error, without considering the consequences on the update center.

From that point on, all autoreleases will necessarily need to fail.

> get rid of the maintainer-specified component altogether in an autoreleased plugin

I expect this will severely limit the acceptance of this system. Not everyone is comfortable with a version number "2412", especially if Autorelease is just being tried out for a given plugin -- no way to get back to version 3.0.x or whatever, after you've released version 2412.

Another problem would be projects like library plugins keeping lockstep with upstream releases, just adding another digit for the Jenkins plugin itself.

Reviewer 2: > From that point on, all autoreleases will necessarily need to fail.

… unless accompanied by a major version increase. This is something that a PR build status could tell maintainers.

----

I think we'll have the best chance by imposing as few unnecessary restrictions as on plugin developers, if we want it to gain acceptance. Making the branch and the version scheme configurable is one of those things that should be doable without crazy workarounds and will make it easier for people to give this a try.

Reviewer 3: > projects like library plugins keeping lockstep with upstream releases, just adding another digit for the Jenkins plugin itself

```xml
<artifactId>apache-httpcomponents-client-4-api</artifactId>
<version>4.5.5.${changelist}</version>
<properties>
  <changelist>999999-SNAPSHOT</changelist>
</properties>
```

producing versions such as `4.5.5.66.b6af896b0883`, which will sort reasonably. Does mean you would need to pick another flag for `incrementals-publisher`.

Reviewer 2: Then the library releases version `4.5.5.1`. What now?

Reviewer 2: What is basically what this proposal comes down to, with an extra layer to ensure going from `1.2` to `1.2.1` will not have unintended consequences.

Reviewer 3: > you would need to pick another flag for `incrementals-publisher`

For the record: a POM property would not work since the flattened POM omits properties, but you could check for the presence/contents of a named file at that commit.

> What is basically what this proposal comes down to

Really? The proposal as currently written sounds very different to me: a special webhook, a special CI server, etc. The approach I am discussing is something I could probably implement in about a day as a single patch to `incrementals-publisher` (plus prototype usage in some minor plugin I control), given the ability to test changes iteratively like I had from @rtyler when it was initially developed.

> an extra layer to ensure going from `1.2` to `1.2.1` will not have unintended consequences

Is that danger actually specific to this JEP or autoreleasing generally? You can already produce shadowed releases with `maven-release-plugin`, admittedly via a somewhat different kind of source edit. Maybe we should consider other ways of dealing with this class of mistake more broadly, such as having `update-center2` produce a warning (delivered somewhere TBD) when it notices a chronologically new release which is not going to actually be visible in any of the configured update centers. (Not _necessarily_ a mistake, since the maintainer might be producing this release for consumption in some other channel, such as an OEM WAR.)

Reviewer 4: @kshultzCB 
This is an example of a discussion that is non-blocking for this first PR.  You'll want to make sure that the salient points are added to the reasoning section, but this is discussion of the design itself not of the structure of the JEP.

Reviewer 3: > you would need to pick another flag for `incrementals-publisher`

Oh! No special flag needed. Only that `-rc` is not present in the version number.

Reviewer 4: ```suggestion
will fail.

[WARNING]
====
*(bitwiseman)* 
There's a long thread here: https://github.com/jenkinsci/jep/pull/244/files#r293534268 that needs to be integrated into this. 

```

3. Validation that no extra fields are present in the file. Comments are allowed, but any unexpected 
lines will cause validation to fail.

If validation of `Autorelease.yaml` fails for any reason, the build is not performed nothing gets deployed, 
and GitHub is notified of the failure.

==== Deployment of a successful release

> When Karl merges a commit into the `master` branch, that merge commit is commit number 150, and has the 

Reviewer 3: ```suggestion
When Karl merges a commit into the `master` branch, that merge commit is link:https://github.com/jenkinsci/jep/blob/master/jep/305/README.adoc#basic-usage[commit number] 150, and has the 
```

Author: These links are super helpful.

SHA `1a2b3c4`. The following takes place:

* A webhook is sent to `${trusted-ci-server}`, and `Autorelease.yaml` is validated.

> * Once validation passes, a build is performed. If the build passes all its tests, a release 

Reviewer 3: This build is happening on trusted.ci.jenkins.io or whatever, in addition to the build happening on ci.jenkins.io? Or is ci.jenkins.io the “trusted CI server”?

Author: Because the build is getting triggered by a webhook set up specifically for this purpose, the CD builds could be built only on _trusted.ci.jenkins.io_, bypassing the building it twice problem.

Reviewer 2: ci.jenkins.io is not involved here at all (except for a CI build of the master branch).

trusted.ci.j.io would be possible, but that instance is a legacy mess.

Would rather go with a shiny new autorelease.ci.j.io on the Jenkins infra VPN.

Reviewer 3: > except for a CI build of the master branch

That is what I just said. You push to `master` and two Jenkins services initiate builds. Or am I missing something?

Reviewer 2: OK, then we agree here.

Reviewer 4: ```suggestion
* Once validation passes, a build is performed. If the build passes all its tests, a release 

WARNING: Some clarification needed based on https://github.com/jenkinsci/jep/pull/244/files#r293536274 .

```

is generated. In our example, that release number would be `3.1.4-150-1a2b3c4`

> * The built plugin gets deployed to Artifactory

Reviewer 3: This already happens in ci.jenkins.io builds. The only thing that does _not_ happen today is that the update center does not pick up incremental builds (from `master`).

Author: OK, cool. We could adapt that same process here.

> * The resulting plugin appears on the Jenkins Update Center, as Karl Plugin version `3.1.4-150-1a2b3c4`.

Reviewer 3: Appears how?

Author: Hopefully, by the same UC generator mechanism (or a similar mechanism) that does this from Artifactory today.

Author: This currently reads:

- Once validation passes, a build is performed. If the build passes all its tests, a release is generated. In our example, that release number would be 3.1.4-150-1a2b3c4
- The built plugin gets deployed to Artifactory
- The UC generator makes the plugin available on the Jenkins Update Center, as Karl Plugin version 3.1.4-150-1a2b3c4.

== Motivation

At present, Jenkins plugins are, typically, not released on a continous basis. They 
are also not released from a single source of truth, such as a trusted Continuous 
Integration server like link:https://ci.jenkins.io[https://ci.jenkins.io]. 

The notion of continuous delivery of plugin releases has been discussed previously <<footnote-1,^(1)^>>.
Considering that Jenkins is a system used to facilitate Continuous Delivery for many users, it makes 
sense -- and builds credibility -- for the Jenkins developer community to adopt this same practice. 

Having a centralized release system made available to plugin maintainers also provides additional 
confidence that security best practices are being followed. 

> Continuous delivery from trusted CI is something which plugin maintainers can opt into, but is 

Reviewer 3: ```suggestion
Continuous delivery from trusted CI is something which plugin maintainers can opt in to, but is 
```

not required. If a plugin maintainer chooses to continue to follow their own path for releasing 
versions of their code, they remain free to do so.

== Reasoning

=== What's with the version numbering?

Jenkins plugin maintainers are already familiar with the way that Incrementals appends a commit 
number, plus a SHA, to version numbers. These mechanically-generated version numbers offer 
the ability to predictably sort them, so that external systems, such as the Jenkins update center, 
can correctly publish the "newest" version. The addition of a merge SHA also allows for at-a-glance 
feedback to tell people which commits went into the release.

WARNING: We know the problem of consistent version numbering, and the resulting ability to 
sort versions, is important. The scheme discussed here is only one of several possible 
choices. We expect some _lively debate_ around this topic. But it's very important to ensure 
that versions can be easily sorted by systems such as the Jenkins update center. Also under 
consideration is defining a fixed number of fields in version numbers, and considering any 
deviation from that a validation failure.

> === Why YAML?

Reviewer 3: Can probably be dropped. I cannot imagine anyone raising an eyebrow over _this_ aspect.

Author: Yeah, maybe you're right.

Reviewer 4: In case some asks why not JSON or Groovy?

YAML is becoming increasingly common in the Jenkins community, for many reasons. YAML is:

* Human readable
* In use by things like the Kubernetes plugin

> * Easily parsed by any number of publily available libraries

Reviewer 3: ```suggestion
* Easily parsed by any number of publicly available libraries
```

=== Aren't we missing some details about infrastructrue requirements?

> In short, yes, we are. This JEP exists to get the conversation started. Once consensus has been reached, 

Reviewer 3: Well, OK, but I find it hard to imagine anyone objecting to the _idea_ of the JEP. Every time the topic is brought up, people ask why we did not have it yesterday. The devil will be in the details of the automation. In particular, the security model needs to be very carefully reviewed.

a separate Infrastructure Enhancement Proposal (IEP) will be created to go along with this JEP. It would 
be premature to describe every detail of implementation before consensus is reached.

== Backwards Compatibility

> Autorelease introduces no new risks with regard to backwards compatibility of plugins themselves. 

Reviewer 3: Except to the extent that plugin maintainers might forget that merging a PR constitutes public release—there is no separate stage of manually validating behavior and thinking about changes before cutting a release.

Author: I'll add some words to this effect. A helpful reminder if nothing else.

I mean, hopefully, the maintainers will know what they're getting themselves into, since they had to take the action of adding `Autorelease.yaml` to their repository. But you never know...

Reviewer 2: I wonder whether an "Autorelease enablement" status check for PRs would be useful here. You see those when you merge.

Alternatively, a possible practice here could be to have a separate 'autorelease' branch – similar to having master/develop, master/autorelease, and releasing from the latter, could be a viable option for people who want to have finer-grained control other releases.

Reviewer 3: IMO people who want finer-grained control over releases can stick with the status quo.

However, there will be a requirement for plugin maintainers to use a consistent version numbering 
scheme. Consistent version numbering will prevent problems with sorting versions. This is discussed 
elsewhere in this document.

With any plugin upgrade, there are backwards compatibility concerns, and Autorelease is no different 
in that regard. Without Autorelease, there is still nothing stopping a plugin maintainer from releasing 
a backwards-breaking change.

Plugin maintainers are also not required to use Autorelease at all. By taking no action, these 
maintainers will see no change at all to the way they do their Jenkins plugin work.

== Security

Autorelease can help to make Jenkins plugins, and their release processes, more secure in a 

> number of ways. Including, but not limited to:

Reviewer 3: ```suggestion
number of ways, including but not limited to:
```

=== Containment of credentials

By using a single system of record for these builds, a service account, maintained by the 

> JENKINS-CERT team, can be used to access GitHub, deployment to Nexus, and deployment to 

Reviewer 3: ```suggestion
Jenkins CERT team, can be used to access GitHub, deployment to Nexus, and deployment to 
```

> JENKINS-CERT team, can be used to access GitHub, deployment to Nexus, and deployment to 

Reviewer 3: ```suggestion
JENKINS-CERT team, can be used to access GitHub, deploy to Nexus, and deploy to 
```

Author: Changed to read as follows:

By using a single system of record for these builds, a service account, maintained by the 
Jenkins CERT team, can be used to access GitHub, deploy to Nexus, and thus deploy to 
the update centers

> the update centers. Plugin maintainers need not leave their own credentials on a CI server 

Reviewer 3: In fact they need not have Artifactory credentials at all.

which they don't own, and permissions already in place in their GitHub repositories provide 
the required controls over who can merge and release.

=== Automatic enforcement of security best practices
Autorelease builds will all come from `${trusted-ci-server}`. Rules 
can be put in place on `${trusted-ci-server}`, which can provide implicit enforcement of the 
Jenkins infrastructure team's security best practices. Compliance to these best practices becomes 
something that plugin maintainers need not worry about.

== Infrastructure Requirements

> We will need a number of things to get this going. The low level technical details will be described 

Reviewer 3: ```suggestion
We will need a number of things to get this going. The low-level technical details will be described 
```

in a separate Infrastructure Enhancement Request, so, this should be considered a summary for now:

> 1. The webhook, configured on each participating plugin . Security implications of this are a bit beyond the scope of this document so far.

Reviewer 3: ```suggestion
1. The webhook, configured on each participating plugin. Security implications of this are a bit beyond the scope of this document so far.
```

> 1. The webhook, configured on each participating plugin . Security implications of this are a bit beyond the scope of this document so far.

Reviewer 3: As noted above, I think this is not a requirement, since you could use an organization-wide webhook.

Author: I think this is addressed via my additions to the Reasoning section, which now under Reasoning --> Webhooks and the Webhook Receiver.

2. A receiver for the aforementioned webhook, because the `${trusted-ci-server}` will be protected 
behind a VPN
3. A trusted Jenkins server for performing builds and deployments. 
4. `buildPlugin` will need code added to validate `Autorelease.yaml` for correctness, and 
build the plugin according to the settings described therein

> 5. Service account(s), managed by the Jenkins Infrastructure team, which provide secure credentials 

Reviewer 3: FWIW in JEP-305 the deployment is done outside Jenkins, so while the team may manage this service account it is not at risk of compromise from Jenkins itself.

Reviewer 2: This Jenkins could be on a nonpublic network and report results back using GH statuses. The challenge for incrementals was the deployment of the same artifacts as on public PR build, this doesn't need to have this limitation.

to systems such as Artifactory

CAUTION: TODO: Things such as service accounts and new servers will be handled in a separate IEP.

> == Testing

Reviewer 3: I think you are misinterpreting this section of the JEP. This paragraph should be moved to **Reasoning**. The **Testing** section should described how the functionality _of the JEP_ will be tested. In this case I assume it would be interactively, not via automation.

Author: You're right on the money. I misunderstood the intent here. Will move things around.

Autorelease brings with it a heightened importance for quality automated tests. However, there will be 
no rules governing this. As is the case today, plugin maintainers are encouraged to release only 
well-tested code, but there is little to stop someone from releasing something which is under-tested. 
Autorelease does not change this in any way.

== Prototype Implementation

As a proof of concept, the (github-branch-source?) plugin will be the first to adopt. This provides 
the initiative with a heavily used plugin, which sees relatively frequent releases already.

A sample fork of github-branch-source could be provided as a reference implementation for 
this proposal. It is understood that this need not be completed before this JEP is 
"link:https://github.com/jenkinsci/jep/tree/master/jep/1#accepted[accepted]", but will need to 
be made available before this JEP is given 
"link:https://github.com/jenkinsci/jep/tree/master/jep/1#final[Final]" status.

== References

> [[footnote-1]]1. Jenkins World 2017, link:http://bit.ly/2x1lCUZ[Contributor Summit Notes], pp. 11-12

Reviewer 3: ```suggestion
[[footnote-1]]1. Jenkins World 2017, link:[Contributor Summit Notes]
```

Author: If I make this change, we lose the ability for readers to find that document:

![image](https://user-images.githubusercontent.com/21689198/59638072-5c4edb00-9125-11e9-8472-0d6471623529.png)

Is that intentional?

Reviewer 3: Did I not include a URL? I meant to. https://docs.google.com/document/d/1JSxYNI_RuA8ITlxVmxBdFg1A-sOKz-w7a9tzuPfWmr4/edit#heading=h.n2d90ci56ugy

//| Resolution
//| :bulb: Link to relevant post in the jenkinsci-dev@ mailing list archives :bulb:

|===

== Abstract

This JEP introduces a new system which enables Jenkins plugin maintainers to add Continuous Delivery (CD) to their projects.
It is:

* Opt-in: plugin maintainers can choose to use this
* Simple to configure: very little needs to be changed in order to enable CD
* Trusted CI/CD: deployments are produced from a single source of truth, which is managed by 
the Jenkins project

== Specification

=== Overview

We'll use a short story to illustrate what it will look like when a plugin maintainer 
named Karl decides that he wants to perform continuous delivery of his (fictional) `karl-plugin`.

To do so, Karl will:

1. Verify support. Karl's plugin is already in the 
link:https://github.com/jenkinsci[jenkinsci Organization on GitHub], and is 
using `buildPlugin` when built on ci.jenkins.io.

2. Choose a branch from which to release. Karl decides he wants to release from the `master` branch, 
as that is the simplest.

3. Settle on a version numbering scheme. Karl decides on link:https://semver.org/[semantic versioning].

> 4. Add a file to the top directory of his repository called `Autorelease.yaml`. 

Reviewer 4: ```suggestion
4. Add a file to the top directory of his repository called `Autorelease.yaml`. 

WARNING: *jvz* Couldn't this be done as an argument to buildPlugin()?
```

2. `versionNumberFields` must match the `revision` property in `pom.xml`. E.g., if `versionNumberFields` is 
set to `3` (three fields), and the `revision` in `pom.xml` is set to `3.1.4.5` (four fields), validation 
will fail.

3. Validation that no extra fields are present in the file. Comments are allowed, but any unexpected 
lines will cause validation to fail.

If validation of `Autorelease.yaml` fails for any reason, the build is not performed, nothing gets deployed, 
and GitHub is notified of the failure.

==== Deployment of a successful release

When Karl merges a commit into the `master` branch, that merge commit is link:https://github.com/jenkinsci/jep/blob/master/jep/305/README.adoc#basic-usage[commit number] 150, and has the 
SHA `1a2b3c4`. The following takes place:

* A webhook is sent from GitHub, to the webhook receiver. That hook is forwarded  to 
the trusted CI server.
* `Autorelease.yaml` is validated.
* Once validation passes, a build is performed. If the build passes all its tests, a release 
is generated. In our example, that release number would be `3.1.4-150-1a2b3c4`
* The built plugin gets deployed to Artifactory

> * link:https://github.com/jenkins-infra/update-center2/blob/master/README.md[The UC generator] makes

Reviewer 5: I firmly believe NOT creating a tag is a bad idea.

Reviewer 3: Can you write up your objections in a form that can be incorporated into **Reasoning**?

Reviewer 4: @jtnord You can provide further feedback after this PR is merged.

Author: I've moved the `WARNING` section and changed the link so that it points [here](https://github.com/jenkinsci/jep/pull/244/files#r302487016). We'll come back to this.

Reviewer 3: The anchor on that URL does not resolve for me, FWIW.

Author: This is pretty odd. I found what's wrong but I've got no idea how it got that way. URL needs to point to `discussion#anchor` rather than `files#anchor`.

Reviewer 5: developers are used to being able to see and work with tags for releases.  it makes any tooling that needs to checkout a release work without changes (if a plugin is auto released or not should not matter, and it allows you to not worry if something was auto-released or not.  TL;DR this is not the way that maven users expect things to work)

WARNING: We know the problem of consistent version numbering, and the resulting ability to 
sort versions, is important. The scheme discussed here is only one of several possible 
choices. We expect some _lively debate_ around this topic. But it's very important to ensure 
that versions can be easily sorted by systems such as the Jenkins update center. Also under 
consideration is defining a fixed number of fields in version numbers, and considering any 
deviation from that a validation failure.

=== Why YAML?
YAML is becoming increasingly common in the Jenkins community, for many reasons. YAML is:

* Human readable
* In use by things like the Kubernetes plugin
* Easily parsed by any number of publicly available libraries

=== Webhooks and the Webhook Receiver
Webhooks will be used to trigger the builds on the trusted CI server. The hooks themselves 
will be configured on a per-repository basis. The reasons for choosing individual hooks, 
as opposed to an organization wide hook, are:

* Security of the build systems. Because this trusted CI server will reside on a non-public 
network, there needs to be a proxy server of sorts, which can accept these webhooks and send 

> them, securely, to the trusted CI server.

Reviewer 3: Why does this mean that per-repo webhooks are needed? I am not following the reasoning.

Author: Yeah, I need to reword this somehow. The need for a receiver, and the configuration of the hooks themselves (single-repo versus org) aren't the same issue.

> * Traceability. An organization-wide webhook would send those hooks to the receiver far more 

Reviewer 3: As a stylistic matter, please use one source text line per sentence or major phrase. Easier to comment on, diff, etc.

often than is actually necessary. In the event that something goes wrong with this process - 
the receiver goes down, there is a network outage, etc. - it will be easier for the Infra 
team to triage problems if they're only looking at hooks from plugins which are intended 

> to be released via CD.

Reviewer 3: But conversely you are adding complexity by needing to check whether each repo maintainer remembered to configure this webhook, and configured it correctly. Seems simpler to have an org hook and simply discard events from nonmatching plugins.

* Flexibility for plugin maintainers. Let's say that Karl maintains two plugins - his archaic 
`old-timey-plugin`, and his fancy new `karl-plugin` which we've already talked about. If 
Karl wants to use CD on karl-plugin, but continue to release `old-timey-plugin` from his laptop 

> like he's always done, he can do so, by only setting the webhook up on `karl-plugin`.

Reviewer 3: Does not make sense to me. You are already proposing to gate autorelease by an explicit flag.

Author: On revisiting this you're right. Was unnecessary and has been deleted.
